content/articles/2019-05-itdepends.Rmarkdown

```diff
@@ -0,0 +1,241 @@
+
+```{r}
+library(tidyverse)
+itdepends::dep_usage("~/p/tidyversedashboard") %>%
+  count(pkg, sort = TRUE)
+```
+
+- show the most often used _function_ from each package
+
+```{r}
+itdepends::dep_usage("~/p/tidyversedashboard") %>%
+  group_by(pkg) %>%
+  count(fun) %>%
+  top_n(1) %>%
+  arrange(desc(n)) %>%
+  head()
+```
+
+## Measure weights with `itdepends::dep_weight()`
+
+`dep_weight()` weighs the supplied packages and returns a tibble of 24
+different metrics that can be used to compare them. Which metric might be the
+most important varies by person, so it is useful to look at the full spectrum.
+
+```{r}
+weights <- itdepends::dep_weight(c("dplyr", "data.table"))
+weights
+```
+
+The measures break down into a few rough categories
+
+- package and repository information (`package`, `funs`, `downloads`, `open_issues`, `last_updated`, `forks`, `stars`)
+- dependency information (`num_users`, `num_dev`, `user_deps`, `dev_deps`)
+- compiled binary sizes (`bin_self`, `bin_user`, `bin_dev`)
+- time to compile dependencies, and the raw times (`install_self`, `install_user`, `install_dev`, `self_timings`, `user_timings`, `dev_timings`)
+- CRAN information (`downloads`, `first_release`, `last_release`, `total_releases`, `releases_last_52`)
+
+The total number of dependencies, binary sizes and installation times give you
+a good idea the magnitude of the dependency load.
+
+Equally important however are the downloads, release information and open
+issues, which tell you how long the package has existed, how many people are
+using it and how many open issues exist. Packages that have existed longer,
+with more users and fewer issues should be the most reliable.
+
+## Visualize proportions with `itdepends::dep_plot()`
+
+There are two functions to plot the relative weights of dependencies.
+
+1. `dep_plot_time()` plots the median compilation time on CRAN's servers.
+
+```{r}
+itdepends::dep_plot_time("dplyr")
+```
+
+From the above plot we see that the installation time of dplyr for users is
+mostly due to compiling dplyr itself. However for development dependencies the
+stringi and RSQLite packages also take significant time to install.
+
+
+2. `dep_plot_size()` plots the pre-compiled binary size, which is often the more
+useful measure, as most users install pre-compiled binaries.
+
+```{r}
+itdepends::dep_plot_size("dplyr")
+```
+
+From this we see the largest binary dependencies are the BH package and Rcpp.
+While for development dependencies the stringi package is heavy, as are
+two data packages, Lahman and nycflights13.
+
+## Assist removal with `itdepends::dep_locate()`
+
+If you do decide that a given dependency is too heavy or is too unstable to
+keep, what are the best practices for removing it?
+
+First and most importantly, write _tests_ for the existing
+behavior. This is an essential step which will greatly reduce the introduction
+of new bugs. Only once you have written the tests should start performing the
+replacement.
```

Comment:
Carefully review the operation of the function that you're replacing?  I want to say something here like: many tidyverse functions appear superficially similar to base equivalents, but they often handle edge cases (that you've possibly never seen) more consistently. But without tying it so strongly to tidyverse.

Reply:
Maybe something like:

If you re-implement a function, you should probably port over its tests, to make sure you're handling edge cases the same way.